bpo-26175: Fix SpooledTemporaryFile IOBase abstract

GFernie opened the pull request:
One would assume that this class implements the IOBase abstract. As the
underlying file-like object is either io.BytesIO, io.StringIO, or a true
file object, this is a reasonable abstract expect and to implement.

Regardless, the behaviour of this class does not change much in the case
of the attribute being missing from the underlying file-like; an
AttributeError is still raised, albeit from one additional frame on the
stack trace.

<!-- issue-number: bpo-26175 -->
https://bugs.python.org/issue26175
<!-- /issue-number -->

GFernie committed:
Fix SpooledTemporaryFile IOBase abstract

One would assume that this class implements the IOBase abstract. As the
underlying file-like object is either io.BytesIO, io.StringIO, or a true
file object, this is a reasonable abstract expect and to implement.

Regardless, the behaviour of this class does not change much in the case
of the attribute being missing from the underlying file-like; an
AttributeError is still raised, albeit from one additional frame on the
stack trace.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

GFernie committed:
Add news entry for bpo-26175 patch

merwok commented on Lib/test/test_tempfile.py:
```diff
@@ -982,6 +982,35 @@ def test_basic(self):
+            self.assertTrue(
+                hasattr(f, attribute),
+                '{} attribute missing'.format(attribute)
+            )
```
Minor: this test could be made simpler/faster with sets: assert that set(dir(f)) is a superset of set(dir(IOBase)) (or if that doesn’t work because IOBase has extra internal attributes, compare with `{'read', 'readable', etc}`).

Could you also add a test with isinstance to validate that the class implements IOBase?

merwok reviewed (changes requested):
Patch looks good!  Note that the bug contains a stronger motivation than the first mssage here: «This was discovered when seeking a SpooledTemporaryFile-backed lzma file.»

merwok committed:
Update 2017-08-30-23-14-17.bpo-26175.GOaj9y.rst

GFernie committed:
Make SpooledTemporaryFile subclass IOBase

GFernie committed:
Refactor tests to more reliably test abstract

GFernie commented:
Thanks for the feedback @merwok 

I've made some changes to more accurately test how _SpooledTemporaryFile_ implements _IOBase_. So, there are now two separate tests:

- Assert that an instance of _SpooledTemporaryFile_ is an instance of _IOBase_. Because _IOBase_ already implements the interface (nearly) this makes it redundant to check for the individual attributes that _SpooledTemporaryFile_ should implement.
- Assert that the attributes which not declared in _IOBase_ but should be implemented by subclasses are implemented by _SpooledTemporaryFile_.

From the _IOBase_ docs: 

> Even though IOBase does not declare read(), readinto(), or write() because their signatures will vary, implementations and clients should consider those methods part of the interface.

I didn't expect the Spanish Inquisition!

pppery commented on Lib/tempfile.py:
```diff
@@ -685,6 +685,9 @@ def __exit__(self, exc, value, tb):
     def __iter__(self):
         return self._file.__iter__()
 
+    def __del__(self):
```
This shouldn't be added: deleting the SpooledTemporaryFile will null out the reference to the underline file, and therefor call its` __del__`

vadmium replied:
I think if you inherit the default `IOBase.__del__` implementation, it will call _close_ and defeat any ResourceWarning that might otherwise be emitted. Perhaps it is better to make `__del__` do nothing, ~~or set it to `object.__del__`~~. [Seems that last option doesn’t exist.]

GFernie replied:
I agree: the underlying file should not be explicitly deleted as this is not expected behaviour. I can imagine a situation where someone is deliberately holding a reference to the underlying file and they would not expect/want it to be closed until their own reference has fallen out of scope. I've changed the method to do nothing and added a docstring to reflect this.

Thanks for your feedback

pitrou replied:
> I agree: the underlying file should not be explicitly deleted as this is not expected behaviour

The doc says: """This function operates exactly as TemporaryFile() does, except [irrelevant differences]."""

And then, about TemporaryFile: """On completion of the context or destruction of the file object the temporary file will be removed from the filesystem."""

So it seems the underlying file _should_ be deleted when the file object disappears.

coleifer replied:
The point is that `IOBase` implements a `__del__` which has some side-effects. Those side effects are not desirable here. Any underlying buffers being wrapped, e.g. the `TemporaryFile` or the `BytesIO`, will be gc'd and handled as they should.

I think this implementation is correct.

GFernie committed:
Return new absolute position from underlying file on seek

Part of file interface:
https://docs.python.org/library/io.html#io.IOBase.seek

GFernie committed:
Do nothing on __del__

We don't want to delete the underlying file explicitly as the expected
behaviour is for the file to be deleted *after* it falls out of scope.

GFernie committed:
Return new file size from underlying file on truncate

Part of file interface:
https://docs.python.org/library/io.html#io.IOBase.truncate

pitrou commented on Lib/test/test_tempfile.py:
```diff
@@ -982,6 +982,22 @@ def test_basic(self):
+        # SpooledTemporaryFile should implement io.IOBase
+        self.assertIsInstance(self.do_create(), io.IOBase)
+
+    def test_iobase_interface(self):
```
Those tests are not very interesting. It would be better to test the methods operate properly.

coleifer replied:
It *literally* delegates everything to the underlying file.

The existing tests should cover the behaviors. The new behaviors are intended to flesh out the API so that `SpooledTemporaryFile`, and these tests seem to cover that.

pitrou reviewed (changes requested):
Here are some comments. You will also need to update the documentation in `Doc/Library/tempfile.rst` (add a `versionchanged` tag documenting the change)

nubirstein reviewed (approved):
I had a case with python3.7 and latest Flask. When trying to send a file via Flask request, I got an error that SpooledTemporaryFile didn't have seekable() method. I ran my code with this tempfile.py implementation and it works perfectly. I hope to see these changes in the official python release very soon.

GFernie committed:
Merge remote-tracking branch 'origin/master' into bpo-26175_fix-SpooledTemporaryFile-IOBase

GFernie committed:
Add versionchanged tag

merwok commented on Doc/library/tempfile.rst:
```diff
@@ -116,6 +116,7 @@ The module defines the following user-callable items:
 
    .. versionchanged:: 3.8
       Added *errors* parameter.
+      Implemented :py:data:`io.IOBase` inteface.
```
```suggestion
      Implemented :class:`io.IOBase` inteface.
```

simpler, no?

GFernie replied:
Yes :slightly_smiling_face: thanks!

GFernie committed:
Use :class: instead of :py:data:

Co-Authored-By: Éric Araujo <merwok@netwok.org>

GFernie committed:
Revert "Do nothing on __del__"

This reverts commit b588dec6a80c7e113ef9fe7e22749914810f4810.

GFernie committed:
Add test for rolled file finalisation

The underlying file should be deleted when the SpooledTemporaryFile
object is deleted. This follows the same behaviour as TemporaryFile and
is consistent with the documentation. `__del__` should have the same
consequences as using `del` or allowing the object to fall out of scope;
an expectation which b588dec (reverted) broke somewhat.

pitrou commented on Doc/library/tempfile.rst:
```diff
@@ -116,6 +116,7 @@ The module defines the following user-callable items:
 
    .. versionchanged:: 3.8
       Added *errors* parameter.
+      Implemented :class:`io.IOBase` inteface.
```
Since 3.8.0 has been released, this should be moved to a new `versionchanged:: 3.9` block.

pitrou commented on Lib/tempfile.py:
```diff
@@ -679,6 +679,9 @@ def __exit__(self, exc, value, tb):
         return self._file.__iter__()
 
+    def __del__(self):
+        return self._file.__del__()
```
Hmm, calling the `__del__` method directly like that is not correct. `__del__` is only to be called when all external references are gone. Instead I would suggest something like the following:
```python
    def __del__(self):
        if not self.closed:
            import warnings
            warnings.warn('unclosed file %r' % (self,), ResourceWarning,
                          stacklevel=2, source=self)
            self.close()
```
(this is adapted from `FileIO.__del__` in `Lib/pyio.py`)

pitrou commented on Lib/test/test_tempfile.py:
```diff
@@ -1000,6 +1016,21 @@ def test_del_on_close(self):
+        # ResourceWarning since the file was not explicitly closed.
+        f = self.do_create(max_size=2)
+        f.write(b'foo')
+        filename = f.name
```
Does this actually work? Here I get something like:
```python
>>> f = tempfile.SpooledTemporaryFile()
>>> f.name
>>> f.rollover()
>>> f.name
3
```

pitrou commented on Lib/test/test_tempfile.py:
```diff
@@ -1000,6 +1016,21 @@ def test_del_on_close(self):
+        f.write(b'foo')
+        filename = f.name
+        self.assertTrue(os.path.exists(filename))
+        with self.assertWarns(ResourceWarning):
```
Do we also expect a `ResourceWarning` to be raised when the file is not rolled over? If yes, there should be a test for it. If not, then the `__del__` implementation can be removed.

pitrou commented on Lib/test/test_tempfile.py:
```diff
@@ -1000,6 +1016,21 @@ def test_del_on_close(self):
+        filename = f.name
+        self.assertTrue(os.path.exists(filename))
+        with self.assertWarns(ResourceWarning):
+            f.__del__()
```
It is not correct to call `__del__` directly. Instead, use `del f` or `f = None`.

pitrou reviewed (changes requested):
Thanks for posting this. Generally speaking, calling `__del__` directly is incorrect. Only the interpreter should do that. From a user's point of view, what's important is whether an object is still reachable or not.

thehesiod commented on Lib/tempfile.py:
```diff
@@ -743,18 +755,21 @@ def tell(self):
         if size is None:
-            self._file.truncate()
+            return self._file.truncate()
         else:
```
this else is not needed, PEP warning, might as well fix

merwok replied:
It's not necessary to remove the else.  It does no harm.

thehesiod replied:
I never said it was necessary, just bad style as it unnecessarily increased the code depth, take my comment as an FYI

danieldjewell replied:
Just wanted to confirm @merwok - nothing needs to be changed here for approval?

merwok replied:
Right, I advise to not change the `else` line.
